File: comparative_tests/go_tests/image_manipulation_functions/blur_filter.go
package image_manipulation_functions

import (
	"fmt"
	bft "go_tests/basic_functions"
	"image"
	"image/color"
	"math"
)

func BlurFilter(img image.Image, blurIntensity int, outputsPath string) {
	// Ensure the output directory exists
	if err := EnsureOutputDir(outputsPath); err != nil {
		fmt.Println(err)
		return
	}

	// Get the image dimensions
	bounds := img.Bounds()
	width, height := bft.GetImageDimensions(img)

	blurredImage := image.NewRGBA(bounds)

	// Create the blur kernel
	kernelSize := 2*blurIntensity + 1
	kernel := make([][]float64, kernelSize)
	for i := range kernel {
		kernel[i] = make([]float64, kernelSize)
		for j := range kernel[i] {
			kernel[i][j] = 1.0 / float64(kernelSize*kernelSize)
		}
	}

	// Apply the blur filter
	for y := blurIntensity; y < height-blurIntensity; y++ {
		for x := blurIntensity; x < width-blurIntensity; x++ {
			var rSum, gSum, bSum float64

			// Apply the kernel
			for ky := -blurIntensity; ky <= blurIntensity; ky++ {
				for kx := -blurIntensity; kx <= blurIntensity; kx++ {
					// Get the neighboring pixel
					pxColor := color.RGBAModel.Convert(img.At(x+kx, y+ky)).(color.RGBA)
					weight := kernel[ky+blurIntensity][kx+blurIntensity]

					// Add the weighted color to the sum
					rSum += float64(pxColor.R) * weight
					gSum += float64(pxColor.G) * weight
					bSum += float64(pxColor.B) * weight
				}
			}

			// Set the new pixel color
			blurredImage.Set(x, y, color.RGBA{
				R: uint8(math.Min(math.Max(rSum, 0), 255)),
				G: uint8(math.Min(math.Max(gSum, 0), 255)),
				B: uint8(math.Min(math.Max(bSum, 0), 255)),
				A: 255,
			})
		}
	}

	// Create the output file
	outputFile, err := CreateOutputFile(outputsPath)
	if err != nil {
		fmt.Println(err)
		return
	}

	// Encode the blurred image as JPEG and save it to the output file
	if err := SaveImageAsJPEG(outputFile, blurredImage); err != nil {
		fmt.Println(err)
		return
	}

	fmt.Println("Blurred image saved to:", outputsPath)
}

File: comparative_tests/python_tests/image_manipulation_functions/blur_filter.py
from PIL import Image
import numpy as np
from . import utils
import time


def blur_filter(img, intensity, outputs_path):
    start_time = time.time()

    # Ensure the output directory exists
    utils.ensure_output_dir(outputs_path)

    # Get the image dimensions
    width, height = img.size

    # Convert image to numpy array
    img_array = np.array(img.convert("RGBA"))

    # Create the blur kernel
    kernel_size = 2 * intensity + 1
    kernel = np.ones((kernel_size, kernel_size)) / (kernel_size * kernel_size)

    # Create the output file
    blurred_array = np.zeros_like(img_array)

    # Apply the blur filter
    for y in range(intensity, height - intensity):
        for x in range(intensity, width - intensity):
            r_sum, g_sum, b_sum = 0, 0, 0

            # Apply the kernel to the image
            for ky in range(-intensity, intensity + 1):
                for kx in range(-intensity, intensity + 1):
                    px = img_array[y + ky, x + kx]
                    weight = kernel[ky + intensity, kx + intensity]

                    r_sum += px[0] * weight
                    g_sum += px[1] * weight
                    b_sum += px[2] * weight

            # Limit the RGB values to 255
            blurred_array[y, x] = [int(r_sum), int(g_sum), int(b_sum), 255]

    # Conversion of the numpy array to an image
    blurred_img = Image.fromarray(blurred_array, "RGBA")

    # Save the blurred image
    output_file = utils.create_output_file(outputs_path)
    utils.save_image_as_jpeg(output_file, blurred_img.convert("RGB"))

    print("Blurred image saved to:", output_file)
    print("Execution time:", time.time() - start_time, "seconds")
